# <center> TP N°2 </center>
### <center> Résolution des systmes linéairess (Méthodes Itératives)
</center>
<div>
    <center> RAZAFINDRAZAKA Henintsoa </center>
    <center> Wang James </center>
    <center> ____________</center>
    <center> ING5 SI GR02 </center>
    <center> 19/10/2020 </center>
</div>

(b) Application :on définit la matrice A∈Mn,n et le vecteur b∈Mn,1 tels que


$$
\mathbf{A}
=
\begin{bmatrix}
3 & 1 & 0 & \cdots & \cdots & \cdots & 0 
\\ 
1 & 3 & 1 & 0 & \cdots & \cdots & 0 
\\
0 & 1 & 3 & 1 & 0 & \cdots & \vdots 
\\
\vdots & \ddots & \ddots & \ddots & \ddots & \ddots & \vdots 
\\
\vdots & \ddots & \ddots & \ddots & \ddots & \ddots & \vdots 
\\
0 & \cdots & \cdots & 0 & 1 & 3 & 1 
\\
0 & \cdots & \cdots & \cdots & 0 & 1 & 3
\end{bmatrix}
$$

$$
\mathbf{B}
=
\begin{bmatrix}
1 \\ \cdots \\ \cdots \\ \cdots \\ 1
\end{bmatrix}
$$

(i) Montrer que la méthode de Gauss-Seidel pour résoudre le système Ax = b va converger. 

On peut voir directement que A est symétrique et défini positive car tous ces mineurs principaux sont supérieurs ou égaux à zéro alors Gauss-Seidel converge. 

On peut également dire que A est une matrice à diagonale dominate car pour tout i appartenant à [1,n] : $$|a_{ii}|  \ge  \sum \limits_{j=1,j \neq i}^{n}|a_{ij}|$$

(ii) On choisit n = 10. Construire la matrice A (on pourra utiliser la commande diag) et le vecteur b.
Déterminer la solution de Ax = b par la méthode de Cholesky, puis par celle de Gauss-Seidel.


In [25]:
import numpy as np

n = 10
A = 3 * np.eye(n) 
for i in range(n-1):
    A[i][i+1] = 1
    A[i+1][i] = 1

print(A)
print('\n|A|= ', np.linalg.det(A))

b = np.ones((n,1))
b

[[3. 1. 0. 0. 0. 0. 0. 0. 0. 0.]
 [1. 3. 1. 0. 0. 0. 0. 0. 0. 0.]
 [0. 1. 3. 1. 0. 0. 0. 0. 0. 0.]
 [0. 0. 1. 3. 1. 0. 0. 0. 0. 0.]
 [0. 0. 0. 1. 3. 1. 0. 0. 0. 0.]
 [0. 0. 0. 0. 1. 3. 1. 0. 0. 0.]
 [0. 0. 0. 0. 0. 1. 3. 1. 0. 0.]
 [0. 0. 0. 0. 0. 0. 1. 3. 1. 0.]
 [0. 0. 0. 0. 0. 0. 0. 1. 3. 1.]
 [0. 0. 0. 0. 0. 0. 0. 0. 1. 3.]]

|A|=  17710.999999999993


array([[1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.]])

##### Méthode de Cholesky:

A est Symétrique définie positive  alors il existe une unique matrice C triangulaire inférieure dont les coefficients de la diagonale principale sont positifs :$$ A = C.C^{t}$$

In [30]:
def cholesky(A):
    C = np.zeros((n,n))
    C[0][0] = np.sqrt(A[0][0])

    for i in range(1, n):
        C[i][0] = A[i][0] / C[0][0]

    for j in range(1, n):
        s0 = 0 
        for k in range(0, j):
            s0 += C[j][k]**2
        C[j][j] = np.sqrt(A[j][j] - s0)
        for i in range (j, n):
            s1 = 0
            for k in range(0, j):
                s1 += C[i][k] * C[j][k]
            C[i][j] = (1 / C[j][j]) * ( A[i][j] - s1) 
    return C

In [32]:
C = cholesky(A)
print(C)

[[1.73205081 0.         0.         0.         0.         0.
  0.         0.         0.         0.        ]
 [0.57735027 1.63299316 0.         0.         0.         0.
  0.         0.         0.         0.        ]
 [0.         0.61237244 1.62018517 0.         0.         0.
  0.         0.         0.         0.        ]
 [0.         0.         0.6172134  1.61834719 0.         0.
  0.         0.         0.         0.        ]
 [0.         0.         0.         0.61791438 1.61807967 0.
  0.         0.         0.         0.        ]
 [0.         0.         0.         0.         0.61801654 1.61804065
  0.         0.         0.         0.        ]
 [0.         0.         0.         0.         0.         0.61803144
  1.61803496 0.         0.         0.        ]
 [0.         0.         0.         0.         0.         0.
  0.61803362 1.61803413 0.         0.        ]
 [0.         0.         0.         0.         0.         0.
  0.         0.61803393 1.61803401 0.        ]
 [0.         0.      

In [34]:
def trianginf(T,b):
    n = len(b)
    x = np.zeros(b.shape)

    x[0] = b[0] / T[0][0]
    for i in range(1,n):
        sigma = 0
        for j in range(0, i):
            sigma += T[i][j] * x[j]
        x[i] = (1/T[i][i]) * ( b[i] - sigma )
    return x

def triangsup(T,b):
    n = len(b)
    x = np.zeros(b.shape)

    x[n-1] = b[n-1] / T[n-1][n-1]
    for i in range(n-2,-1,-1):
        sigma = 0
        for j in range(i+1, n):
            sigma += T[i][j] * x[j]
        x[i] = (1/T[i][i]) * ( b[i] - sigma )
    return x

In [36]:
y = trianginf(C,b) 
print("y : \n ",y)

x = triangsup(C.T,y)
print("\nx: \n", x)
print("\nnp.linalg.inv(A).dot(b)) : \n", np.linalg.inv(A).dot(b))

print("\nb : \n", b)
print("\nA.dot(x) : ", A.dot(x))

y : 
  [[0.57735027]
 [0.40824829]
 [0.46291005]
 [0.44136741]
 [0.44946657]
 [0.44635604]
 [0.44754158]
 [0.44708838]
 [0.44726143]
 [0.44719532]]

x: 
 [[0.27638191]
 [0.17085427]
 [0.21105528]
 [0.1959799 ]
 [0.20100503]
 [0.20100503]
 [0.1959799 ]
 [0.21105528]
 [0.17085427]
 [0.27638191]]

np.linalg.inv(A).dot(b)) : 
 [[0.27638191]
 [0.17085427]
 [0.21105528]
 [0.1959799 ]
 [0.20100503]
 [0.20100503]
 [0.1959799 ]
 [0.21105528]
 [0.17085427]
 [0.27638191]]

b : 
 [[1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]]

A.dot(x) :  [[1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]]


##### Méthode de Gauss-Seidel:

In [54]:
print("\nA : \n", A)
D = 3 * np.eye(n) 
M , N = np.zeros((n,n)), np.zeros((n,n))

for i in range(n-1):
    N[i][i+1] = -1
    M[i+1][i] = -1
    
print("\nD : \n", D)
print("\nM : \n", M)
print("\nN : \n", N)


A : 
 [[3. 1. 0. 0. 0. 0. 0. 0. 0. 0.]
 [1. 3. 1. 0. 0. 0. 0. 0. 0. 0.]
 [0. 1. 3. 1. 0. 0. 0. 0. 0. 0.]
 [0. 0. 1. 3. 1. 0. 0. 0. 0. 0.]
 [0. 0. 0. 1. 3. 1. 0. 0. 0. 0.]
 [0. 0. 0. 0. 1. 3. 1. 0. 0. 0.]
 [0. 0. 0. 0. 0. 1. 3. 1. 0. 0.]
 [0. 0. 0. 0. 0. 0. 1. 3. 1. 0.]
 [0. 0. 0. 0. 0. 0. 0. 1. 3. 1.]
 [0. 0. 0. 0. 0. 0. 0. 0. 1. 3.]]

D : 
 [[3. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 3. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 3. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 3. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 3. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 3. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 3. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 3. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 3. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 3.]]

M : 
 [[ 0.  0.  0.  0.  0.  0.  0.  0.  0.  0.]
 [-1.  0.  0.  0.  0.  0.  0.  0.  0.  0.]
 [ 0. -1.  0.  0.  0.  0.  0.  0.  0.  0.]
 [ 0.  0. -1.  0.  0.  0.  0.  0.  0.  0.]
 [ 0.  0.  0. -1.  0.  0.  0.  0.  0.  0.]
 [ 0.  0.  0.  0. -1.  0.  0.  0.  0.  0.]
 [ 0.  0.  0.  0.  0. -1.  0.  0.  0.  0.]
 [ 0.  0.  0.  0

In [82]:
NbIterationMax = 9
epsilon = 0.001
X_old = np.zeros((n,1))
X_new = np.zeros((n,1))
epochs = 0

for i in range(n):
    sn, sm = 0, 0
    for j in range(i+1, n):
        sn += N[i][j] * X_old[j]
    for j in range(0, i):
        sm += M[i][j] * X_new[j]
    X_new[i] = (1/D[i][i]) * (sn + sm + b[i])

test_darret = np.linalg.norm(X_new - X_old)**2 / np.linalg.norm(X_new)**2

while (test_darret >= epsilon) or (epochs < NbIterationMax):
    epochs += 1
    X_old = X_new
    
    for i in range(n):
        sn, sm = 0, 0
        for j in range(i+1, n):
            sn += N[i][j] * X_old[j]
        for j in range(0, i):
            sm += M[i][j] * X_new[j]
        X_new[i] = (1/D[i][i]) * (sn + sm + b[i])
        
    test_darret = np.linalg.norm(X_new - X_old)**2 / np.linalg.norm(X_new)**2
    print("nb of epochs : ", epochs)

if i > NbIterationMax: 
    print("Gauss-Seidel n’a pas convergé")
else:
    print("Gauss-Seidel a convergé avec une précision de ", epsilon)
    
    print("\nx : \n", X_new)

nb of epochs :  1
nb of epochs :  2
nb of epochs :  3
nb of epochs :  4
nb of epochs :  5
nb of epochs :  6
nb of epochs :  7
nb of epochs :  8
nb of epochs :  9
Gauss-Seidel a convergé avec une précision de  0.001

x : 
 [[0.27639256]
 [0.17083339]
 [0.21108257]
 [0.19595182]
 [0.2010292 ]
 [0.20098706]
 [0.19599155]
 [0.21104871]
 [0.17085735]
 [0.27638088]]


 X trouvé avec la méthode de Cholesky : 
 [[0.27638191]
 [0.17085427]
 [0.21105528]
 [0.1959799 ]
 [0.20100503]
 [0.20100503]
 [0.1959799 ]
 [0.21105528]
 [0.17085427]
 [0.27638191]]
 
 Gauss Seidel converge a partir de 10 itérations pour un précision de 0,1. 
 